source/extensions/tracers/opentelemetry/samplers/dynatrace/sampler_config_provider.cc

```diff
@@ -8,10 +10,98 @@ namespace Extensions {
+  // we want to ensure that we don't have more than one pending request
+  if (timeout > (TIMER_INTERVAL / 2)) {
+    timeout = (TIMER_INTERVAL / 2);
+  }
+  return timeout;
+}
+
+} // namespace
+
 SamplerConfigProviderImpl::SamplerConfigProviderImpl(
-    Server::Configuration::TracerFactoryContext& /*context*/,
+    Server::Configuration::TracerFactoryContext& context,
     const envoy::extensions::tracers::opentelemetry::samplers::v3::DynatraceSamplerConfig& config)
-    : sampler_config_(config.root_spans_per_minute()) {}
+    : cluster_manager_(context.serverFactoryContext().clusterManager()),
+      http_uri_(config.http_uri()),
+      authorization_header_value_(absl::StrCat("Api-Token ", config.token())),
+      sampler_config_(config.root_spans_per_minute()), timeout_(getTimeout(http_uri_)) {
+
+  timer_ = context.serverFactoryContext().mainThreadDispatcher().createTimer([this]() -> void {
+    const auto thread_local_cluster = cluster_manager_.getThreadLocalCluster(http_uri_.cluster());
+    if (thread_local_cluster == nullptr) {
+      ENVOY_LOG(error, "SamplerConfigProviderImpl failed: [cluster = {}] is not configured",
+                http_uri_.cluster());
+    } else {
+      Http::RequestMessagePtr message = Http::Utility::prepareHeaders(http_uri_);
+      message->headers().setReferenceMethod(Http::Headers::get().MethodValues.Get);
+      message->headers().setReference(Http::CustomHeaders::get().Authorization,
+                                      authorization_header_value_);
+      active_request_ = thread_local_cluster->httpAsyncClient().send(
+          std::move(message), *this, Http::AsyncClient::RequestOptions().setTimeout(timeout_));
+    }
+  });
+
+  timer_->enableTimer(std::chrono::seconds(INITIAL_TIMER_DURATION));
+}
+
+SamplerConfigProviderImpl::~SamplerConfigProviderImpl() {
+  if (active_request_) {
+    active_request_->cancel();
+  }
+}
+
+void SamplerConfigProviderImpl::onSuccess(const Http::AsyncClient::Request& /*request*/,
+                                          Http::ResponseMessagePtr&& http_response) {
+  active_request_ = nullptr;
+  const auto response_code = Http::Utility::getResponseStatus(http_response->headers());
+  bool json_valid = false;
+  if (response_code == enumToInt(Http::Code::OK)) {
+    ENVOY_LOG(debug, "Received sampling configuration from Dynatrace: {}",
+              http_response->bodyAsString());
+    json_valid = sampler_config_.parse(http_response->bodyAsString());
+    if (!json_valid) {
+      ENVOY_LOG(warn, "Failed to parse sampling configuration received from Dynatrace: {}",
+                http_response->bodyAsString());
+    }
+  } else {
+    ENVOY_LOG(warn, "Failed to get sampling configuration from Dynatrace: {}", response_code);
+  }
+
+  if (json_valid || reEnableTimer(static_cast<Http::Code>(response_code))) {
+    timer_->enableTimer(std::chrono::seconds(TIMER_INTERVAL));
+  }
```

Comment:
Is it reasonable design not to enable the timer if the body is invalid or the response code is unexpected value?

And I think at least here should have an error log to tell user the timer will never be enabled again.

Reply:
Response codes lie 400, 401. 403, 404, ... are caused by a config error (wrong uri, wrong/invalid token). Therefore we thought the re-enabling the timer does not make sense.
I agree, we should log an error if we never enable the timer.